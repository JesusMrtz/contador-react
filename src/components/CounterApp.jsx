import { useState } from 'react';
import PropTypes from 'prop-types';


export function CounterApp({ value }) {
    console.log('Renderizar nuevamente');
    
    /** Utilizando el hook useState con valor inicial del valor */
    const [ counter, setCounter ] = useState(value);

    function handleIncrease(e) {
        /**
         * Cambia el estado del "counter" para renderizar el HTML donde se imprime el "counter"
        */
        setCounter( counter + 1 );
    }

    function handleDecrease() {
        setCounter(counter - 1);
    }
    
    function handleReset() {
        setCounter(value);
    }

    
    return(
        <>
            <h1>CounterApp</h1>
            <h2>{ counter }</h2>
            <hr />
            <button type='button' onClick={ handleIncrease }>+1</button>
            <button type='button' onClick={ handleDecrease }>-1</button>
            <button type='button' onClick={ handleReset } >Reset</button>
        </>
    );
}


/** Creaci√≥n de los propTypes  */
CounterApp.propTypes = {
    value: PropTypes.number.isRequired
}