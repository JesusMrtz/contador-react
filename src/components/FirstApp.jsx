import PropTypes  from 'prop-types';


/**
 * Si no hay una dependencia de comportamiento del componente, las variables y funciones pueden estar fuera
 * Aquellas funciones o variables que estan fuera del Functional Componet, no se vuelve a redibuajar en memoria para mostrarlo.
 * 
 */

const name = 'Jesús Martínez Torres';
const age = 26;
const arrayOfNumbers = [1, 2, 3, 4, 5, 6, 7, 8, 9];
const newMessage = {
    author: name,
    message: 'Hola, soy goku'
};

function returnValue() {
    return 'Estoy retornando algo';
}


function FirstApp({ title, propNumber }) {
    return (
        <>
            <h1>{ title }</h1>
            <h2>Estoy recibiendo un prop númerico: { propNumber * 12}</h2>
            <hr />
            <h2>Hola, mi nombre es { name }</h2>
            <h3>{ arrayOfNumbers }</h3>

            {/* Si queremos imprimir un objeto, tenemos que serealizarlo */}
            <p>{ JSON.stringify(newMessage) }</p>
            {/*  */}

            <p>{ returnValue() }</p>

            <h3>Este es un subtitulo del subtitulo { age * 43 }</h3>
        </>
    );
}

/** Agregando tipado a los PropTypes  */
FirstApp.propTypes = {
    title: PropTypes.string.isRequired,
    propNumber: PropTypes.number
}

/** Implementando defaultProps  */
FirstApp.defaultProps = {
    title: 'No hay titulo',
    // propNumber: 0
}


export default FirstApp;