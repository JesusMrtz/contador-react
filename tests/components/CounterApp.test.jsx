import { fireEvent, render, screen } from "@testing-library/react";
import { CounterApp } from "../../src/components/CounterApp";

describe(`Pruebas en el componente <CounterApp />`, () => {
    const value = 100;

    it('Debe hacer match con el snapshot', () => {
        const { container } = render(<CounterApp value={ value }  />);
        expect(container).toMatchSnapshot();
    });


    it('Debe mostrar el valor inicial de 100', () => {
        render(<CounterApp  value={ value } />);
        expect( screen.getByRole('heading', { level: 2 }).textContent ).toContain( String(value) );
    });


    it('Incrementar el valor inicial a 102', async () => {
        render(<CounterApp value={ value } />);
        const getButton = screen.getByText('+1');

        await fireEvent.click(getButton);
        await fireEvent.click(getButton);

        expect( screen.getByText('102') ).toBeTruthy();
    });


    it('Decrementar el valor inicial a 98', async () => {
        render(<CounterApp value={ value } />);
        const getButton = screen.getByText('-1');

        await fireEvent.click(getButton);
        await fireEvent.click(getButton);

        expect( screen.getByText('98') ).toBeTruthy();
    });


    it('Regresar el valor inicial a 1000', async () => {
        render(<CounterApp value={ value } />);
        const getButton = screen.getByText('-1');
        const buttonReset = screen.getByText('Reset');

        await fireEvent.click(getButton);
        await fireEvent.click(getButton);
        await fireEvent.click(buttonReset);

        expect( screen.getByText('100') ).toBeTruthy();
    });
});