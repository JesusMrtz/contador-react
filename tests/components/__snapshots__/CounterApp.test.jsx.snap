// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <CounterApp /> Debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2>
    100
  </h2>
  <hr />
  <button
    type="button"
  >
    +1
  </button>
  <button
    type="button"
  >
    -1
  </button>
  <button
    type="button"
  >
    Reset
  </button>
</div>
`;
